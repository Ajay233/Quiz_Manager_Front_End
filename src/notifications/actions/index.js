export const setNotification = (msg, type, show) => {
  return {
    type: "SET_NOTIFICATION",
    payload: {
      message: msg,
      type: type,
      show: show
    }
  }
}
