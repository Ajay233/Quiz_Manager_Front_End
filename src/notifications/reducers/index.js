import { notification } from '../factory/notificationFactory'

export const setNotificationReducer = (notificationData=notification, action) => {
  if(action.type === "SET_NOTIFICATION"){
    return action.payload;
  } else {
    return notification;
  }
}
