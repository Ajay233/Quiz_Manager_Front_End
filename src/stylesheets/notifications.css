.error {
  border: 2px #f5c6cb solid;
  background-color: #f8d7da;
  padding: 15px;
  font-style: normal;
  color: #721c24;
  border-radius: 5px;
}

.success {
  border: 2px #c3e6cb solid;
  background-color: #d4edda;
  padding: 15px;
  font-style: normal;
  color: #155724;
  border-radius: 5px;
}

.warning {
  border: 2px #ffeeba solid;
  background-color: #fff3cd;
  padding: 15px;
  font-style: normal;
  color: #856404;
  border-radius: 5px;
}
