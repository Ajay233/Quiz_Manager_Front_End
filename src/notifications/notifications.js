import React from 'react'
import '../stylesheets/notifications.css'
import '../stylesheets/buttons.css'
import { connect } from 'react-redux';
import { setNotification } from './actions'

import { timedFunc } from '../utils/display'
import { isIconRequired, iconPicker } from '../utils/icons'

class Notification extends React.Component {

  componentDidMount(){
    this.notificationRef = React.createRef();
    this.imgRef = React.createRef();
  }

  // take a parameter to determine an icon needs to be rendered
  // decide if timer needs to be applied
  // decide what icon to rende
  // make a close button

  renderMsg = () => {
    const { setNotification } = this.props;
    const { show, type, timed } = this.props.notificationData;
    if(show){
      if (timed) timedFunc(3470, setNotification);
      document.documentElement.scrollTop = 0;
      return isIconRequired(type) ? this.renderMsgWithIcon(type, timed) : this.renderStandardMsg(type)
    } else {
      return null;
    }
  }

  renderMsgWithIcon = (type, timed) => {
    return(
      <div
        ref={this.notificationRef}
        id="notification"
        className={`spacing ${type}${timed ? "shrinkNotification" : ""}`}
      >
        {timed ? null : <span className="close" onClick={this.handleClose}><i className="far fa-times-circle"></i></span>}
        <div className="notificationBody">
          <div className="notificationImg">
            <img ref={this.imgRef} src={require(`../public/icons/${iconPicker(type)}`)} className={`img ${timed ? "shrinkImg" : ""}`} alt=""/>
          </div>
          <div className="notificationText">
            {this.props.notificationData.message}
          </div>
        </div>
      </div>
    );
  }

  renderStandardMsg = (type) => {
    return(
      <div
        ref={this.notificationRef}
        id="notification"
        className={`spacing shrinkNotification ${type}`}
      >
        {this.props.notificationData.message}
      </div>
    );
  }

  handleClose = () => {
    const { setNotification } = this.props;
    this.notificationRef.current.classList.add("shrinkNotificationNow")
    this.imgRef.current.classList.add("shrinkImgNow")
    window.setTimeout(function(){
      setNotification();
    }, 487)
  }

  render(){
    return(
      <React.Fragment>
        {this.renderMsg()}
      </React.Fragment>
    );
  }
}

export const mapStateToProps = (state) => {
  return {
    notificationData: state.notificationData
  }
}

export default connect(mapStateToProps, { setNotification })(Notification)
