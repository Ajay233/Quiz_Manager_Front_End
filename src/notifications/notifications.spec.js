import React from 'react'
import { Provider } from 'react-redux'
import Notification from './notifications'
import configureStore from 'redux-mock-store'
import renderer from 'react-test-renderer'
import { mapStateToProps } from './notifications'

const mockStore = configureStore([]);

describe('mapStateToProps', () => {
  it('can map state to props', ()=> {
    const notificationTestData = {
      message: "My test notification",
      type: "success",
      show: true
    }
    const appState = {
      userData: {},
      verificationProcess: {},
      redirect: {},
      notificationData: notificationTestData,
      listOfUsers: {},
      form: {}
    }
    const componentState = mapStateToProps(appState);
    expect(componentState.notificationData).toEqual(notificationTestData);
  })
});

describe('Notification component', () => {

  it('can render a notification', () => {

    let store;
    let component;

    store = mockStore({
      notificationData: {
        message: "Test message",
        type: "success",
        show: true,
        timed: true
      }
    });

    component = renderer.create(
      <Provider store={store}>
        <Notification />
      </Provider>
    );

    expect(component).toMatchSnapshot();

  })

  it('can hide a notification message when the show property is set to false', () => {
    let store;
    let component;

    store = mockStore({
      notificationData: {
        message: "",
        type: "",
        show: false,
        timed: true
      }
    });

    component = renderer.create(
      <Provider store={store}>
        <Notification />
      </Provider>
    );

    expect(component).toMatchSnapshot();
  })

});
