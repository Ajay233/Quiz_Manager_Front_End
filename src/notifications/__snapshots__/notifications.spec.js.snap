// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Notification component can hide a notification message when the show property is set to false 1`] = `
<div
  className="spacing"
  id="notification"
/>
`;

exports[`Notification component can render a notification 1`] = `
<div
  className="spacing"
  id="notification"
>
  <div
    className="success"
  >
    <span
      className="close"
      onClick={[Function]}
    >
      <i
        className="far fa-times-circle"
      />
    </span>
    Test message
  </div>
</div>
`;

exports[`Notification component can render a notification with an icon 1`] = `
<div
  className="spacing"
  id="notification"
>
  <div
    className="verifyProcess"
  >
    <span
      className="close"
      onClick={[Function]}
    >
      <i
        className="far fa-times-circle"
      />
    </span>
    <span>
      <span>
        <img
          alt=""
          className="img"
          src="mail-sent.png"
        />
      </span>
      test of notification that includes an icon png
    </span>
  </div>
</div>
`;
